### PR TITLE
feat: add Alertmanager's status subresource

### PR DESCRIPTION
**Pull request by simonpasquier**
## Description

Following #4580, this change implements the status subresource for the Alertmanager CRD.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [X] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Add Alertmanager's status subresource
```

### CODE REVIEW

File: test/e2e/prometheus_test.go [hunk 1]
```diff
@@ -4507,6 +4507,66 @@ func testPromQueryLogFile(t *testing.T) {
...
+	framework.SetupPrometheusRBAC(context.Background(), t, testCtx, ns)
+
+	p := framework.MakeBasicPrometheus(ns, "test", "", 2)
+	// A non-existing service account prevents the creation of the statefulset's pods.
+	p.Spec.ServiceAccountName = "does-not-exist"
+
+	_, err := framework.MonClientV1.Prometheuses(ns).Create(context.Background(), p, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (JoaoBraveCoding) [inline test/e2e/prometheus_test.go:4524]:
```suggestion
	require.NoError(t, err)
```

File: test/framework/alertmanager.go [hunk 2]
```diff
@@ -231,24 +249,61 @@ func (f *Framework) WaitForAlertmanagerReady(ctx context.Context, ns string, a *
...
 		ctx,
-		ns,
-		5*time.Minute,
-		int(*a.Spec.Replicas),
-		alertmanager.ListOptions(a.Name),
+		name,
+		types.ApplyPatchType,
+		b,
+		metav1.PatchOptions{
+			Force:        func(b bool) *bool { return &b }(true),
```
Comment (JoaoBraveCoding) [inline test/framework/alertmanager.go:286]: Can't we use `"k8s.io/utils/pointer"` here?
```suggestion
			Force:        pointer.Bool(true),
```

File: test/e2e/alertmanager_test.go [hunk 3]
```diff
@@ -1839,17 +1839,33 @@ func testAMRollbackManualChanges(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	replicas := int32(0)
-	sset.Spec.Replicas = &replicas
-	if _, err := ssetClient.Update(context.Background(), sset, metav1.UpdateOptions{}); err != nil {
+	sset.Spec.Replicas = func(i int32) *int32 { return &i }(0)
```
Comment (JoaoBraveCoding) [inline test/e2e/alertmanager_test.go:1842]: nit: can't we use "k8s.io/utils/pointer" here?
```suggestion
	sset.Spec.Replicas = pointer.Int32(0)
```

**Review by JoaoBraveCoding — COMMENTED** [review]
just some minor comments, but apart from that looks awsome 💯

File: pkg/alertmanager/operator.go [hunk 4]
```diff
@@ -726,9 +766,133 @@ func (c *Operator) sync(ctx context.Context, key string) error {
...
+
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	a := pobj.(*monitoringv1.Alertmanager)
+	a = a.DeepCopy()
```
[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/alertmanager/operator.go:783]: This is to ensure that the object won't be modified underneath or that the function will not modify it further down the line. I can't find the documentation stating that it's a good practice but it's the safest approach.

File: pkg/alertmanager/operator.go [hunk 5]
```diff
@@ -726,9 +766,133 @@ func (c *Operator) sync(ctx context.Context, key string) error {
...
+	if err != nil {
+		return errors.Wrap(err, "failed to retrieve statefulset state")
+	}
+
+	aStatus.Replicas = int32(len(stsReporter.Pods))
+	aStatus.UpdatedReplicas = int32(len(stsReporter.UpdatedPods()))
+	aStatus.AvailableReplicas = int32(len(stsReporter.ReadyPods()))
+	aStatus.UnavailableReplicas = int32(len(stsReporter.Pods) - len(stsReporter.ReadyPods()))
+
+	if len(stsReporter.ReadyPods()) < replicas {
```
[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/alertmanager/operator.go:837]: good catch!

[the remark this answers is not in the document]
Comment (JoaoBraveCoding) [reply pkg/alertmanager/operator.go:837]: TIL

File: test/framework/alertmanager.go [hunk 6]
```diff
@@ -231,24 +249,61 @@ func (f *Framework) WaitForAlertmanagerReady(ctx context.Context, ns string, a *
...
+	return p, nil
+}
+
+func (f *Framework) ScaleAlertmanagerAndWaitUntilReady(ctx context.Context, name, ns string, replicas int32) (*monitoringv1.Alertmanager, error) {
+	return f.PatchAlertmanagerAndWaitUntilReady(
+		ctx,
+		name,
+		ns,
+		monitoringv1.AlertmanagerSpec{
+			Replicas: func(i int32) *int32 { return &i }(replicas),
```
Comment (JoaoBraveCoding) [inline test/framework/alertmanager.go:304]: nit
```suggestion
			Replicas: pointer.Int32(replicas)
```

**Review by JoaoBraveCoding — COMMENTED** [review]
just one final that slipped through the cracks but from my side after that I'm fine with giving it ✅

**Review by PhilipGough — APPROVED** [review]
@simonpasquier thanks for doing this.

@JoaoBraveCoding thanks for the great job on the prior reviews as I was a bit late to the party here

**Pull request merged**